### PR TITLE
Get rb_num2uint and rb_num2ulong working.

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: truffleruby/src/main/c/cext/ruby.c [hunk 1]
```diff
@@ -379,15 +379,15 @@ int NUM2INT(VALUE value) {
 }
 
 unsigned int NUM2UINT(VALUE value) {
-  return (unsigned int) truffle_invoke_i(RUBY_CEXT, "NUM2UINT", value);
+  return (unsigned int) truffle_invoke_l(RUBY_CEXT, "NUM2LONG", value);
```
Comment (chrisseaton) [inline truffleruby/src/main/c/cext/ruby.c]: Why this change?

File: truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java [hunk 2]
```diff
@@ -211,9 +214,20 @@ public long long2num(long num) {
...
+                // The l at the end of the constant is crucial,
+                // otherwise the constant will be interpreted as an
+                // int (-1) and then converted to a long (still -1,
+                // 0xffffffffffffffff), and have completely the wrong
+                // effect.
+                BigInteger lsi = BigInteger.valueOf(num & 0xffffffffl);
+                BigInteger msi = BigInteger.valueOf(num >>> 32).shiftLeft(32);
+                BigInteger res = msi.add(lsi);
+                return BignumOperations.createBignum(getContext(), res);
+            }
```
Comment (chrisseaton) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: We probably need a boundary on this.

File: truffleruby/src/main/c/cext/ruby.c (hunk 1 above)
Comment (aardvark179) [inline truffleruby/src/main/c/cext/ruby.c]: NUM2UINT is effectively a truncation of NUM2LONG, and must handle conversion of numbers that don't fit into a signed Java int.

We should probably revisit these CextNodes anyway as ruby defines these to be either macros wrapping the rb_num2xxx variants, or inlined versions of them so the current CextNodes aren't complete in their behaviour. I don't think there are any specs for most of these though.

File: truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java (hunk 2 above)
Comment (aardvark179) [inline truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java]: Added.

File: truffleruby/src/main/java/org/truffleruby/core/CoreLibrary.java [hunk 3]
```diff
@@ -1099,6 +1099,10 @@ public static boolean fitsIntoInteger(long value) {
         return ((int) value) == value;
     }
 
+    public static boolean fitsIntoUnsignedInteger(long value) {
+        return value == (value & 0xffffffffl) || value < 0 && value >= Integer.MIN_VALUE;
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/core/CoreLibrary.java]: This seems a bit strange of a condition. Should a small negative long "fit" in a unsigned integer?
Could the condition be reduced to just 2 comparisons?

File: truffleruby/src/main/c/cext/ruby.c [hunk 4]
```diff
@@ -379,15 +379,15 @@ int NUM2INT(VALUE value) {
...
 }
 
 long NUM2LONG(VALUE value) {
   return truffle_invoke_l(RUBY_CEXT, "NUM2LONG", value);
 }
 
 unsigned long rb_num2ulong(VALUE val) {
-  rb_tr_error("rb_num2ulong not implemented");
+  return (unsigned long)truffle_invoke_l(RUBY_CEXT, "rb_num2ulong", val);
```
Comment (aardvark179) [inline truffleruby/src/main/c/cext/ruby.c]: We pass back long from Java containing the appropriate bits, and the explicit cat doesn't change it. All a bit sneaky as Java doesn't have unsigned types for us to use.

File: truffleruby/src/main/java/org/truffleruby/core/CoreLibrary.java (hunk 3 above)
Comment (aardvark179) [inline truffleruby/src/main/java/org/truffleruby/core/CoreLibrary.java]: So, I would hope nothing after the `||` is ever relevant, but the rule in MRI appears to be as follows
1. Anything which can fir into a signed int (positive or negative) can be turned into an unsigned int
2. Anything integer which can be represented in 32 bits can be turned into an unsigned int.

So a small negative long can be cast to an unsigned int because it can be simply truncated to a signed int without loss of meaning and then cast to be unsigned. This function is not normally passed small negative longs in our implementation, but I put the check in there so it would give the correct answer even if we slipped up elsewhere.

Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/core/CoreLibrary.java]: I think it's best to keep our internals clean. If we don't need it at some caller, then call a normal implementation. Only use this extended range when needed.
And still, why these obscure conditions over `Integer.MIN_VALUE <= value && value < (1L << 64)`?
